### PR TITLE
[Enhancement] use heuristic method to handle left deep join

### PR DESCRIPTION
**Pull request by dirtysalt**
## What type of PR is this：
- [ ] bug
- [ ] feature
- [X] enhancement
- [ ] refactor
- [ ] others

## Which issues of this PR fixes ：

Fixes #

## Problem Summary(Required) ：

The main idea is to build inner join first. If there is no inner join, then fallback to cross join.

use tpcds-1g query64 as test case. thers is no column stats, only table row count information:
- without this PR, there are 4 cross joins. and estimated rows are 4930002698541411 (max rows are 12449931856073464)
- with this PR, there are no cross joins, and estimated rows are 914616 (max rows are 2333127)

### CODE REVIEW

**Comment by dirtysalt** [comment]
> UT plz

Yeah, I'll add UT later. Just to see if it works for you first.

**Comment by dirtysalt** [comment]
run starrocks_fe_unittest

**Comment by dirtysalt** [comment]
run starrocks_fe_unittest

**Comment by wanpengfei-git** [comment]
#### [FE PR Coverage Check]

:heart_eyes: **pass** : 32 / 32 (100.00%)




#### file detail

|   |path|covered_line|new_line|coverage|not_covered_line_detail|
|----|----|----|----|----|----|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java|18|18|100.00%|[]|
|:large_blue_circle:|com/starrocks/sql/optimizer/rule/join/JoinOrder.java|14|14|100.00%|[]|

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java [hunk 1]
```diff
@@ -24,8 +25,35 @@ protected void enumerate() {
             return (diff < 0 ? -1 : (diff > 0 ? 1 : 0));
         });
 
+        boolean useHeuristic = true;
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java:28]: Seems this var is unnecessary?

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java [hunk 2]
```diff
@@ -24,8 +25,35 @@ protected void enumerate() {
...
-        for (int index = 1; index < atomSize; ++index) {
+        used[0] = true;
+        int next = 1;
+        while (next < atomSize) {
+            if (used[next]) {
+                next++;
+                continue;
+            }
+            int index = next;
+            Preconditions.checkState(!used[index]);
```
Comment (liuyehcf) [inline fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java:39]: Seems this check is unnecessary

**Review by Seaven — APPROVED** [review]

**Review by Youngwb — APPROVED** [review]

**Comment by wanpengfei-git** [comment]
run starrocks_admit_test

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java (hunk 2 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java:39]: defensive programming:)

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java (hunk 1 above)
Comment (dirtysalt) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java:28]: Yes. I put it here for easier debugging: if you doubt this strategy is better and want to fallback to old strategy.

**Review by HangyuanLiu — APPROVED** [review]

File: fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java (hunk 2 above)
Comment (liuyehcf) [reply fe/fe-core/src/main/java/com/starrocks/sql/optimizer/rule/join/JoinReorderLeftDeep.java:39]: Emm, that's not defensive, it is sure that used[index] must be false

**Review by liuyehcf — APPROVED** [review]

**Pull request merged**